drivers/st25r3911b/nfca.c

```diff
@@ -0,0 +1,1244 @@
+					     &nfca.water_lvl.rx);
+	if (err) {
+		return err;
+	}
+
+	irq = ST25R3911B_IRQ_MASK_RXS | ST25R3911B_IRQ_MASK_RXE |
+	      ST25R3911B_IRQ_MASK_TXE | ST25R3911B_IRQ_MASK_COL |
+	      ST25R3911B_IRQ_MASK_TIM | ST25R3911B_IRQ_MASK_ERR |
+	      ST25R3911B_IRQ_MASK_CRC | ST25R3911B_IRQ_MASK_PAR |
+	      ST25R3911B_IRQ_MASK_ERR2 | ST25R3911B_IRQ_MASK_ERR1 |
+	      ST25R3911B_IRQ_MASK_FWL;
+
+	if (fdt != 0) {
+		irq |= ST25R3911B_IRQ_MASK_NRE | ST25R3911B_IRQ_MASK_GPE;
+
+		err = transfer_fdt_set(fdt);
+		if (err) {
+			return err;
+		}
+	}
+
+	err = transmission_prepare();
+	if (err) {
+		return err;
+	}
+
+
+	/* Enable TX and RX interrupts */
+	err = st25r3911b_irq_enable(irq);
+	if (err) {
+		return err;
+	}
+
+	/* Write Tx data to FIFO, if data is longer than FIFO write,
+	 * then only write first 96 bytes.
+	 */
+	err = st25r3911b_fifo_write(tx->data,
+				    MIN(tx->len, ST25R3911B_MAX_FIFO_LEN));
+	if (err) {
+		return err;
+	}
+
+	/* Set Rx buffer */
+	nfca.transfer.rx_buf = rx;
+
+	/* Set data length to send */
+	err = st25r3911b_tx_len_set(tx->len);
+	if (err) {
+		return err;
+	}
+
+	if (auto_crc) {
+		cmd = ST25R3911B_CMD_TX_WITH_CRC;
+
+		LOG_DBG("Transmit data with auto generated CRC");
+
+	} else {
+		cmd = ST25R3911B_CMD_TX_WITHOUT_CRC;
+
+		LOG_DBG("Transmit data without CRC");
+	}
+
+	nfca.transfer.written_byte = MIN(tx->len, ST25R3911B_MAX_FIFO_LEN);
+
+	return st25r3911b_cmd_execute(cmd);
+}
+
+static int tag_sleep(void)
+{
+	int err;
+	u8_t tx_buffer[] = NFCA_SLEEP_CMD;
+	struct nfca_buf tx = {
+		.data = tx_buffer,
+		.len = sizeof(tx_buffer)
+	};
+
+	err = transfer(&tx, NULL, 0, true);
+	if (err) {
+		return err;
+	}
+
+	nfca.tag.sleep = true;
+
+	return 0;
+}
+
+static int user_state_process(void)
+{
+	int err = 0;
+
+	switch (atomic_get(&nfca.state.tag)) {
+	case STATE_FIELD_ON:
+		err = st25r3911b_field_on(ST25R3911B_NO_THRESHOLD_ANTICOLLISION,
+					  ST25R3911B_NO_THRESHOLD_ANTICOLLISION,
+					  NFCA_GT_TIME);
+
+		state_set(STATE_IDLE);
+
+		if (!err) {
+			if (nfca.cb->field_on) {
+				nfca.cb->field_on();
+			}
+		}
+
+		break;
+
+	case STATE_FIELD_OFF:
+		err = st25r3911b_rx_tx_disable();
+
+		state_set(STATE_IDLE);
+
+		if (!err) {
+			if (nfca.cb->field_off) {
+				nfca.cb->field_off();
+			}
+		}
+
+		break;
+
+	case STATE_TAG_DETECTION:
+		err = tag_detect(nfca.cmd);
+		break;
+
+	case STATE_ANTICOLLISION:
+		err = anticollision_resolution(ANTICOLLISION_CASCADE_1);
+		break;
+
+	case STATE_TRANSFER:
+		err = transfer(nfca.transfer.tx_buf,
+			       nfca.transfer.rx_buf,
+			       nfca.transfer.fdt,
+			       nfca.transfer.auto_crc);
+		break;
+
+	case STATE_SLEEP:
+		err = tag_sleep();
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+static void timeout_handler(struct k_work *work)
+{
+	timeout_process();
+}
+
+void st25r3911b_nfca_process(void)
+{
+	int err = 0;
+
+	k_sem_take(&init_sem, K_FOREVER);
+
+	while (!err) {
+		err = k_poll(events, NFCA_EVENT_CNT, K_FOREVER);
```

Comment:
Few weeks ago we had a chat about it. 
Some samples use the architecture where you have to periodically call process function. I found this approach externally limiting as you won't be able to extend the application with any new functionality without reworking libraries, doing stupid hacks or wasting memory on additional threads.

The easy solution would be to allow application to fetch pollable objects from the library and poll on them. Example of usage:
```
static struct k_poll_event event[4];

void main(void)
{
   st25r3911b_nfca_init();
   st25r3911b_nfca_get_events(&event[0], 2);
   other_lib_init();
   other_lib_get_events(&event[2], 2);
   while (true) {
      k_poll(event, 4, K_FOREVER);

      /* don't check event state here, make lib check it internally.
       * if nothing is available it means processing is not needed for this cycle. */
      st25r3911b_nfca_process();

      other_lib_process();
   }
}
```

For user convenience you could add an optional thread or a function that would do the loop. However library would not limit application.

Reply:
Yes. Good idea!

Reply:
Hi @grochu , @lemrey , @carlescufi , regarding this comment I think it would be good to use the same approach in other our libraries so users could integrate several of them in one thread.

Reply:
Hi, yes, this makes sense. Perhaps we should add a relevant information to our development guidelines, so that other components follow the same scheme.